### DISCUSSION TITLE
Can I avoid the XRButton ?

### DISCUSSION THREAD

**Discussion by quertenmont** [pmndrs/xr#220] [Q&A] [answered]
Hello,

I am looking to develop an XR app for the meta store through a PWA+WEB XR.
But I'd like to skip the step where the user has to push the XR Button to get into the immersive view.

Is this possible or is it a limitation of web xr ?

Thanks for helping,
Loic

**Comment by CodyJasonBennett** [comment, marked as answer]
WebXR requires interaction to start a session, so you could place a button beforehand (you'll often see immersive sites have a button gate to get the same permission for audio), and eagerly create a session once granted. Maybe react-xr can export a method to do this. This would be the relevant code:

```jsx
function XRSessionInit() {
  const set = useXR(state => state.set)

  useLayoutEffect(() => {
    let session

    try {
      navigator.xr
        .requestSession('immersive-vr', {
          optionalFeatures: ['local-floor', 'bounded-floor', 'hand-tracking', 'layers'],
        })
        .then((session) => {
          // Check if unmounted
          if (session === null) return
          // Otherwise, init react-xr
          set(() => ({ session }))
        })
    } catch(e) {
      // Session unsupported
      console.error(e)
    }
    
    return () => {
      session?.end()
      session = null
      set(() => ({ session }))
    }
  }, [set])
}

{hasInteraction && (
  <Canvas>
    <XR>
      <XRSessionInit />
    </XR>
  </Canvas>
)}
```

**Reply by quertenmont** [reply]
Thanks for answering

**Reply by bbohlender** [reply]
As a side node, there's also the `sessiongranted` event, which can be emitted to trusted sites to enter AR/VR without interaction. react-three/xr supports and enables this by default.
https://github.com/immersive-web/navigation
